azurelinuxagent/common/utils/shellutil.py

```diff
@@ -110,6 +110,34 @@ def run_get_output(cmd, chk_err=True, log_cmd=True, expected_errors=[]):
     return 0, output
 
 
+def _encode_command_output(output):
+    return ustr(output, encoding='utf-8', errors="backslashreplace")
+
+
+def run_command(command):
+    """
+    Wrapper for subprocess.Popen.
+    Executes the given command and returns its stdout. Logs any errors executing the command and raises an exception.
+    """
+    try:
+        process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
+        stdout, stderr = process.communicate()
+        retcode = process.returncode
+
+        if retcode:
+            logger.error(u"Command: [{0}], return code: [{1}], "
+                         u"stdout: [{2}] stderr: [{3}]".format(command,
+                                                               retcode,
+                                                               _encode_command_output(stdout),
+                                                               _encode_command_output(stderr)))
+            raise subprocess.CalledProcessError(retcode, command)
```

Comment:
Not sure we should raise this as a CalledProcessError... the documentation is specific about this exception being raised by cehck_call/check_output (vs. Popen:
> a exception subprocess.CalledProcessError
Exception raised when a process run by check_call() or check_output() returns a non-zero exit status.

Using this exception might not be a big issue is that the output and stderr properties would be populated, I think that is missing from your code.

Consider whether the caller actually needs the return code/stdout/stderr of the command and, if not, you might be OK just raising an Exception

Reply:
If we're redesigning run_get_output we should return all 3 - exit_code, out, err and let it be for the caller to decide what to use

Reply:
Removed the raise of CalledProcessError since it's not needed as long as we log on error and raise an exception on failure. Plus, `stderr` was only introduced to the CalledProcessError constructor in Python 3 so trying to avoid supporting both.

@larohra I'm deciding to keep it as returning stdout only for this iteration since we don't consume exit_code and stderr. We will change the design as needed when we make the final switch from using `run_get_output` to `run_command`.